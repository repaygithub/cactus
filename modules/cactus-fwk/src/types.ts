// TS Types only, allows for re-export
// see: https://github.com/babel/babel/issues/8361#issuecomment-435623593
export interface FeatureFlagsObject {
  [k: string]: boolean
}

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>
