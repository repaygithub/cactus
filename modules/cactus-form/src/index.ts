export { default as DependentField } from './DependentField'
export { default as FieldSpy } from './FieldSpy'
export { default as Field, validateRequired } from './Field'
