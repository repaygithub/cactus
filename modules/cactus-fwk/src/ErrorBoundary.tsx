import React, { ComponentType, ErrorInfo, Fragment } from 'react'

export type ErrorView = ComponentType<ErrorViewProps>
export type OnError = (error: Error, info: ErrorInfo) => void

interface ErrorViewProps {
  error: Error
  info: ErrorInfo
}

interface ErrorBoundaryProps {
  onError: OnError
  errorView?: ErrorView
}

interface ErrorBoundaryState {
  error: Error | undefined
  errorInfo: ErrorInfo | undefined
}

const initialState: ErrorBoundaryState = { error: undefined, errorInfo: undefined }

export class ErrorBoundary extends React.Component<ErrorBoundaryProps, ErrorBoundaryState> {
  readonly state: ErrorBoundaryState = initialState

  componentDidCatch(error: Error | null, info: ErrorInfo | null) {
    const { onError } = this.props
    if (error === null) {
      error = new Error('Error was lost during propagation')
    }
    if (info === null) {
      info = { componentStack: 'Stack trace unavailable.' }
    }
    if (typeof onError === 'function') {
      onError(error, info)
    }
    this.setState({ error: error, errorInfo: info })
  }

  render() {
    const { error, errorInfo } = this.state
    const { children, errorView: ErrorView } = this.props

    if (error && errorInfo) {
      return ErrorView ? <ErrorView error={error} info={errorInfo} /> : null
    } else {
      return <Fragment>{children}</Fragment>
    }
  }
}

const noop = (error: Error, info: React.ErrorInfo) => {}

const withErrorBoundary = <BaseProps extends any>(
  BaseComponent: ComponentType<BaseProps>,
  onError?: OnError,
  errorView?: ErrorView
) => {
  const Wrapped = (props: BaseProps) => (
    <ErrorBoundary onError={onError ? onError : noop} errorView={errorView}>
      <BaseComponent {...props} />
    </ErrorBoundary>
  )

  return Wrapped
}

export default withErrorBoundary
