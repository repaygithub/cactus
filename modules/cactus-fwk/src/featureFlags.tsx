import React, { createContext, useContext, ComponentType, FC } from 'react'
import { FeatureFlagsObject } from './types'

const FeatureFlagContext = createContext<FeatureFlagsObject | null>(null)

const useFeatureFlags = <FeatureFlags extends string[]>(...features: FeatureFlags): boolean[] => {
  const featureFlags = useContext(FeatureFlagContext)
  if (featureFlags === null) {
    return Array(features.length).fill(false)
  }
  const result: boolean[] = []
  for (const key of features) {
    result.push(Boolean(featureFlags[key]))
  }
  return result
}

const withFeatureFlags = <FeatureFlags extends string[], P extends object>(
  features: FeatureFlags,
  Component: ComponentType<P>
) => {
  return (props: P) => (
    <FeatureFlagContext.Consumer>
      {featureFlags => {
        const flags: FeatureFlagsObject = {}
        if (featureFlags === null) {
          for (const key of features) {
            flags[key] = false
          }
          return <Component {...props} {...flags} />
        }
        for (const key of features) {
          flags[key] = featureFlags[key]
        }
        return <Component {...props} {...flags} />
      }}
    </FeatureFlagContext.Consumer>
  )
}

interface FeatureFlagProps {
  feature: string
  children: (enabled: boolean) => React.ReactElement
}

const FeatureFlag: FC<FeatureFlagProps> = ({ feature, children }) => {
  const [enabled] = useFeatureFlags(feature)
  return children(enabled)
}

export { FeatureFlagContext, useFeatureFlags, withFeatureFlags, FeatureFlag }
