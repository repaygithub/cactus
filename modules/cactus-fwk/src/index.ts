export { useFeatureFlags, withFeatureFlags, FeatureFlag } from './featureFlags'
export { default } from './AppRoot'

export * from './types'
