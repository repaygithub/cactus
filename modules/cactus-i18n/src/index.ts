export {
  I18nProvider as default,
  I18nSection,
  I18nElement,
  I18nFormatted,
  I18nResource,
  I18nText,
} from './Components'
export { useI18nText, useI18nResource, useI18nSection, I18nContext } from './hooks'
export { default as BaseI18nController } from './BaseI18nController'
export * from './types'
