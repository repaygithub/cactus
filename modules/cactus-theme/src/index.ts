import defaultTheme from './theme'
export * from './helpers/index'
export * from './theme'
export default defaultTheme
