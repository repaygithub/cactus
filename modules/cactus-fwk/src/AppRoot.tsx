import React, { WeakValidationMap } from 'react'

import { ErrorBoundary, ErrorView, OnError } from './ErrorBoundary'
import { FeatureFlagContext } from './featureFlags'
import { FeatureFlagsObject } from './types'
import PropTypes from 'prop-types'

interface AppRootProps {
  /**
   * Receives an object for featureFlags
   */
  featureFlags?: FeatureFlagsObject
  /**
   * Receives an error handler function
   */
  onError?: OnError
  /**
   * Receives a view for global errors
   */
  globalErrorView?: ErrorView
}

const noop = (error: Error, info: React.ErrorInfo) => {}

const AppRoot: React.FC<AppRootProps> = props => {
  return (
    <ErrorBoundary onError={props.onError || noop} errorView={props.globalErrorView}>
      <FeatureFlagContext.Provider value={props.featureFlags || null}>
        <React.Fragment>{props.children}</React.Fragment>
      </FeatureFlagContext.Provider>
    </ErrorBoundary>
  )
}

AppRoot.propTypes = {
  featureFlags: PropTypes.objectOf(PropTypes.bool.isRequired),
  onError: PropTypes.func,
  globalErrorView: PropTypes.element,
} as WeakValidationMap<AppRootProps>

export default AppRoot
