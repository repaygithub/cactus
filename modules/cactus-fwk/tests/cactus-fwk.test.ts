import * as cactusFwk from '../src/index'

describe('@repay/cactus-fwk', () => {
  test('should match snapshot of exports', () => {
    expect(Object.keys(cactusFwk)).toMatchSnapshot()
  })
})
