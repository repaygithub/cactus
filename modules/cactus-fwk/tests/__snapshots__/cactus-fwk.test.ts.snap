// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@repay/cactus-fwk should match snapshot of exports 1`] = `
Array [
  "useFeatureFlags",
  "withFeatureFlags",
  "FeatureFlag",
  "default",
  "withErrorBoundary",
  "ErrorBoundary",
]
`;
